File: 0165/solution.py
class Solution:
    def compareVersion(self, version1: str, version2: str) -> int:
        i, j = 0, 0
        n, m = len(version1), len(version2)

        while i < n or j < m:
            num1 = 0
            while i < n and version1[i] != ".":
                num1 = num1 * 10 + int(version1[i])
                i += 1
            i += 1

            num2 = 0
            while j < m and version2[j] != ".":
                num2 = num2 * 10 + int(version2[j])
                j += 1
            j += 1

            if num1 > num2:
                return 1
            if num1 < num2:
                return -1

        return 0

File: 0165/solution.java
class Solution {
    public int compareVersion(String version1, String version2) {
        int V1 = version1.length();
        int V2 = version2.length();
        int i = 0;
        int j = 0;
        while (i < V1 || j < V2) {
            int num1 = 0;
            while (i < V1 && version1.charAt(i) != '.') {
                num1 = num1 * 10 + (int) (version1.charAt(i++) - '0');
            }
            i++;
            int num2 = 0;
            while (j < V2 && version2.charAt(j) != '.') {
                num2 = num2 * 10 + (int) (version2.charAt(j++) - '0');
            }
            j++;

            if (num1 > num2) {
                return 1;
            }

            if (num2 > num1) {
                return -1;
            }
        }
        return 0;
    }
}

// time O(m + n)
// space O(1)